src/main/resources/db/migration/0300/V333__ObservationPlotBiomassDetails.sql

```diff
@@ -0,0 +1,187 @@
+                OR (species_id IS NULL AND species_name IS NOT NULL)),
+
+    CONSTRAINT invasive_or_threatened
+        CHECK ((is_invasive IS true OR is_threatened IS true)),
+
+    FOREIGN KEY (observation_id, monitoring_plot_id)
+        REFERENCES tracking.observation_biomass_details (observation_id, monitoring_plot_id)
+        ON DELETE CASCADE
+);
+
+CREATE INDEX ON tracking.observation_biomass_additional_species(monitoring_plot_id);
+
+CREATE UNIQUE INDEX ON tracking.observation_biomass_additional_species
+    (observation_id, monitoring_plot_id, species_id)
+    WHERE species_id IS NOT NULL;
+
+CREATE UNIQUE INDEX ON tracking.observation_biomass_additional_species
+    (observation_id, monitoring_plot_id, species_name)
+    WHERE species_name IS NOT NULL;
+
+CREATE TABLE tracking.tree_growth_forms(
+    id INTEGER PRIMARY KEY,
+    name TEXT NOT NULL UNIQUE
+);
+
+CREATE TABLE tracking.recorded_trees (
+    id BIGINT PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
+    observation_id BIGINT NOT NULL REFERENCES tracking.observations ON DELETE CASCADE,
+    monitoring_plot_id BIGINT NOT NULL REFERENCES tracking.monitoring_plots ON DELETE CASCADE,
+    species_id BIGINT REFERENCES species,
+    species_name TEXT,
+    tree_number BIGINT NOT NULL CHECK ( tree_number >= 1 ),
+    tree_growth_form_id INTEGER NOT NULL REFERENCES tracking.tree_growth_forms,
+    is_dead BOOLEAN NOT NULL,
+    is_trunk BOOLEAN,
+    diameter_at_breast_height_cm NUMERIC CHECK ( diameter_at_breast_height_cm >= 0 ),
+    point_of_measurement_m NUMERIC CHECK ( point_of_measurement_m >= 0 ),
+    height_m NUMERIC CHECK ( height_m >= 0 ),
+    shrub_diameter_cm NUMERIC CHECK ( shrub_diameter_cm >= 0 ),
+    description TEXT,
+
+    CONSTRAINT species_identifier
+     CHECK ((species_id IS NOT NULL AND species_name IS NULL)
+         OR (species_id IS NULL AND species_name IS NOT NULL)),
+
+    CONSTRAINT growth_form_specific_data
+        CHECK (
+            (tree_growth_form_id = 1
+                AND is_trunk IS NOT NULL
+                AND diameter_at_breast_height_cm IS NOT NULL
+                AND point_of_measurement_m IS NOT NULL
+                AND shrub_diameter_cm IS NULL) OR
+            (tree_growth_form_id = 2
+                AND is_trunk IS NULL
+                AND diameter_at_breast_height_cm IS NULL
+                AND point_of_measurement_m IS NULL
+                AND height_m IS NULL
+                AND shrub_diameter_cm IS NOT NULL)
+            ),
+
+    CONSTRAINT height_required_dbm_threshold
+        CHECK ( NOT (diameter_at_breast_height_cm > 5 AND height_m IS NULL)),
+
+    FOREIGN KEY (observation_id, monitoring_plot_id)
+        REFERENCES tracking.observation_biomass_details (observation_id, monitoring_plot_id)
+        ON DELETE CASCADE,
+
+    UNIQUE (observation_id, tree_number)
+);
+
+CREATE INDEX ON tracking.recorded_trees(observation_id);
+CREATE INDEX ON tracking.recorded_trees(monitoring_plot_id);
```

Comment:
Don't need the index on observation ID since it's covered by the unique index on `(observation_id, tree_number)`.